### ISSUE TITLE
NPE, cause not known

### ISSUE THREAD

**Issue by j-baker** [immutables/immutables#261]
Hey,
After adding a few interfaces, I've a couple of times been blocked by Immutables NPEing when ./gradlew assemble is run. The stack trace is at the bottom. It doesn't happen when Immutables is run inside of IntelliJ.

I've not been able to find the root cause of this - specifically classes that cause Immutables to NPE in one project compile fine when copied to another project.

    error: org.immutables.value.internal.$processor$.$Processor threw java.lang.NullPointerException
  	at org.immutables.value.internal.$guava$.base.$Preconditions.checkNotNull($Preconditions.java:210)
  	at org.immutables.value.internal.$processor$.meta.$StyleMirror.<init>($StyleMirror.java:793)
  	at org.immutables.value.internal.$processor$.meta.$StyleMirror.from($StyleMirror.java:92)
  	at org.immutables.value.internal.$processor$.meta.$Proto$Environment.defaultStyles($Proto.java:148)
  	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$Environment.<init>($ImmutableProto.java:2034)
  	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$Environment.of($ImmutableProto.java:2272)
  	at org.immutables.value.internal.$processor$.meta.$Round.environment($Round.java:79)
  	at org.immutables.value.internal.$processor$.meta.$ImmutableRound.environment($ImmutableRound.java:288)
  	at org.immutables.value.internal.$processor$.meta.$Round.declaringTypeFrom($Round.java:169)
  	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collectIncludedAndDefinedBy($Round.java:240)
  	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collect($Round.java:190)
  	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collect($Round.java:180)
  	at org.immutables.value.internal.$processor$.meta.$Round.collectProtoclasses($Round.java:129)
  	at org.immutables.value.internal.$processor$.meta.$Round.collectValues($Round.java:83)
  	at org.immutables.value.internal.$processor$.$Processor.process($Processor.java:48)
  	at org.immutables.value.internal.$generator$.$AbstractGenerator.process($AbstractGenerator.java:89)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
  	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
  	at com.sun.tools.javac.main.Main.compile(Main.java:523)
  	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
  	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)

**Comment by elucash** [comment]
Thank you for reporting this! The only hypothesis I have is the gradle is running annotation processing in a separate classpath (which is expected) but actual lookup of elements happens on the compile classpath.
You might need to add "org.immutables:value:2.1.10:annotations" dependency to either "provided" or "compile" scope. Let us know if this will work for you. We might also need to diagnose this better.

**Comment by j-baker** [comment]
Ok, that solved the problem (adding immutables annotations as a compile dependency).

could perhaps immutables simply not use those annotations if the annotations package isn't on the classpath (or otherwise not fail)? A nice property of immutables is that we and downstream consumers don't need to take on an immutables dependency.

**Comment by elucash** [comment]
I agree on the motivation, it is possible to change the code to not read the annotation element to get default values, but I guess the compilation will fail anyway.
I might be wrong in something, but you cannot compile code annotated with the annotations without having these annotations in the classpath. The trick is that (as in our case) you don't need to have those annotations in runtime classpath, nor any client code which uses immutable and abstract value classes have any dependency on the annotations or the processor. Ideal scopes for such annotation dependencies are scopes that are not propagated to runtime or downstream modules, it could be
'compileOnly'  of some sort. Maven doesn't have such scope, so I usually use 'provided' scope and/or 'optional' flag. Gradle, probably have 'provided', and some plugins bring 'compileOnly' scope or equivalent.
Speaking about this, there are two common ways to run annotation processing. The first way is to have it in the compile classpath (along with annotations) of code to compile. For this simple usage which works by default in maven and javac, we have combined artifact `org.immutables:value` which contains annotations and repackaged processor with dependencies. The second way is to run annotation processor separately, providing isolated annotation processing classpath for a processor to run. But in this case, in order to compile code actual code annotations (not processor) still has to be present in the compilation classpath as any other classes that need to be typechecked if used. For this we have this annotation-only artifact with 'annotation' classifier.
Using `android-apt` plugin, the example _Immutables_ artifacts would be

```
dependencies {
  // ...
  // for annotation processor which have it's own classpath
  apt "org.immutables:value:2.1.10"
  
  // annotation-only artifact is still needed for compilation
  // should not propagate to downstream dependencies or runtime
  provided "org.immutables:value:2.1.10:annotations"
}
```

As far as I know other annotation processors such as dagger2 also work around by having annotation artifact and compiler(processor) artifact.

**Comment by j-baker** [comment]
We're using https://github.com/palantir/gradle-processors. It marks the annotation processor as a provided dependency, and sets up IDEA, Eclipse, javadoc etc to use the generator and generated classes appropriately.

I'm quite confused by this bug - to be clear Immutables works fine for most of the cases where we use it (indicating that we're set up correctly in general) - but there are certain classes where their addition will cause this NPE.

As another thing, this does not happen in immutables 2.0.x (but we need 2.1.x because the Jackson codegen handles @JsonAnyGetter and @JsonTypeName properly).

**Comment by j-baker** [comment]
Ok, problem solved. One of our dependencies marked immutables 2.0.21 as a compile dependency and exported it to us.

Possibly worth changing the error message that is presented?

**Comment by elucash** [comment]
sure
newly released v2.1.11 added this as 
```java
processing().getMessager()
 +            .printMessage(Diagnostic.Kind.ERROR,
 +                "Could not found annotations on the compile classpath. It looks like annotation processor is running"
 +                    + " in a separate annotation-processing classpath and unable to get to annotation definitions."
 +                    + " To fix this, please add annotation-only artifact 'org.immutables:value:(version):annotations'"
 +                    + " to 'compile' 'compileOnly' or 'provided' dependency scope.");
```
any ideas how to improve the wording are welcome

**Comment by elucash** [comment]
Released in 2.1.12 with the latest refinements. Thank you!
